### PR TITLE
3213 coldp type materials

### PR DESCRIPTION
**Pull request by gdower**
Adds the TypeMaterial table to CoL data package exports

### CODE REVIEW

File: lib/export/coldp/files/type_material.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+  def self.locality(co)
+    [co.collecting_event&.cached_level0_geographic_name, co.collecting_event&.cached_level1_geographic_name, co.collecting_event&.cached_level2_geographic_name].compact.join(', ')
+  end
+
+  # TODO: This very likely includes more than just hosts (symbiotypes):
+  #   could possibly be narrowed down by checking if the biological relationship type's inverted name includes the word host?
+  #     we are specifying the relationship type in the string to make it clearer
+  def self.host(co)
+    if co.biological_associations == []
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:31]: Use `.empty?` or `.any?` (with not).

File: lib/export/coldp/files/type_material.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+# catalogNumber
+# associatedSequences
+# sex
+# link
+# remarks
+#
+module Export::Coldp::Files::TypeMaterial
+
+  def self.locality(co)
+    [co.collecting_event&.cached_level0_geographic_name, co.collecting_event&.cached_level1_geographic_name, co.collecting_event&.cached_level2_geographic_name].compact.join(', ')
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:24]: There is a method to get all levels and their ranks IIRC.

File: lib/export/coldp/files/type_material.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+  # TODO: This very likely includes more than just hosts (symbiotypes):
+  #   could possibly be narrowed down by checking if the biological relationship type's inverted name includes the word host?
+  #     we are specifying the relationship type in the string to make it clearer
+  def self.host(co)
+    if co.biological_associations == []
+      return nil
+    end
+    hosts = []
+    co.biological_associations.each do |ba|
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:35]: You should add the 'Otu' constraint here if that's what you want, like:

`co.biological_association.where(biological_association_subject_type: 'Otu')`

File: lib/export/coldp/files/type_material.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      if ba.biological_association_object_type == 'Otu'
+        o = Otu.find(ba.biological_association_object_id)
+        if o.taxon_name_id.nil?
+          object_name = o.name
+        else
+          tn = TaxonName.find(Otu.find(ba.biological_association_object_id).taxon_name_id)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:41]: `Otu.find(ba.biological_association_object_id)`

^ You already have the object in o, you don't need another DB call, just `o.taxon_name_id`?

File: lib/export/coldp/files/type_material.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+          object_name = tn.cached
+        end
+        rel_type = BiologicalRelationship.find(ba.biological_relationship_id)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:44]: just `ba.biological_relationship`.  All has_many, has_one, belongs_to give you that magic.

File: lib/export/coldp/files/type_material.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        end
+      end
+    end
+    hosts.to_sentence
+  end
+
+  # CollectingEvent.verbatim_date seems to be in different formats (e.g., "6.VIII.2018", "14 August 1992")
+  # so preferably take start_date_xxxx and format to ISO 8601
+  def self.date(co)
+    date = [co.collecting_event&.start_date_year, co.collecting_event&.start_date_month, co.collecting_event&.start_date_day].compact.join('-')
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:56]: There are date methods you should use in CE.

File: lib/export/coldp/files/type_material.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    if co.collecting_event&.start_date_year.nil? or date.nil?  # if year is nil, verbatim_date might be better?
+      date = co.collecting_event&.verbatim_date
+    end
+    date
+  end
+
+  def self.collector(co)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:63]: You should use the DwC methods likely, see the inlucde for CollectionObjects.  Something like co.dwc_<field_in_snake_case>.  Many, many options there.

File: lib/export/coldp/files/type_material.rb [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    end
+    co.collecting_event.roles.each do |role|
+      collectors.append(Person.find(role.person_id).cached)
+    end
+    collectors.to_sentence
+  end
+
+  # TODO: Should we want to return the current repository acronym if co.current_repository_id is set?
+  #       Will the catalogue number still be the same even if a collection object was moved to another institution?
+  def self.institution_code(co)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:76]: See DwC methods.

**Review by mjy — COMMENTED** [review]
@gdower this looks like a nice structure.  I think there are some things to simplify, see comments.  if at all possible use `co.dwc_<foo>` methods.

File: lib/export/coldp/files/type_material.rb (hunk 2 above)
Comment (gdower) [reply lib/export/coldp/files/type_material.rb:24]: I haven't found a method, but I changed it to:

[co.dwc_country, co.dwc_state_province, co.dwc_county, co.dwc_locality].compact.join(', ')

File: lib/export/coldp/files/type_material.rb [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    end
+    hosts = []
+    co.biological_associations.where(biological_association_object_type: 'Otu').each do |ba|
+      if o.taxon_name_id.nil?
+        object_name = o.name
+      else
+        tn = TaxonName.find(o.taxon_name_id)
+        object_name = tn.cached
+      end
+      if ba.biological_relationship.inverted_name.include? 'host'
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:42]: @gdower I think this wont' really work, or catch only very few posibilities.  I'd suggest leaving out host for now?

File: lib/export/coldp/files/type_material.rb [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        hosts.append("#{ba.biological_relationship.name} #{object_name}")
+      end
+    end
+    hosts.to_sentence
+  end
+
+  # CollectingEvent.verbatim_date seems to be in different formats (e.g., "6.VIII.2018", "14 August 1992")
+  # so preferably take dwc_event_date and format to ISO 8601
+  def self.date(co)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:51]: Should likely use dwc_event_date.

File: lib/export/coldp/files/type_material.rb [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+      date = co.collecting_event&.verbatim_date
+    end
+    date
+  end
+
+  # TODO: is the CollectionObject.preferred_catalog_number the right method to use?
+  #   some CollectionObject identifier types like Identifier::Global::Uuid::TaxonworksDwcOccurrence and
+  #   Identifier::Unknown are not exported by the preferred_catalog_number method, but perhaps that's desirable.
+  #   Alternatively use co.dwc_catalog_number?
+  def self.catalog_number(co)
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:63]: dwc_catalog_number might keep it more in sync

File: lib/export/coldp/files/type_material.rb [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+          csv << [
+            nil,                                                # ID: don't expose TW internal type material ID
+            tm.protonym_id,                                     # nameID
+            co.buffered_collecting_event,                       # citation
+            tm.type_type,                                       # status
+            reference_id,                                       # referenceID
+            locality(co),                                       # locality
+            co.collecting_event&.cached_level0_geographic_name, # country
+            co.collecting_event&.verbatim_latitude,             # latitude
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:108]: dwc_decimal_latitude

File: lib/export/coldp/files/type_material.rb [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            co.collecting_event&.verbatim_longitude,            # longitude
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:109]: Use dwc_ getter for this too. dwc_verbatim_latitude

File: lib/export/coldp/files/type_material.rb [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+            co.collecting_event&.verbatim_elevation,            # altitude
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:110]: dwc_verbatim_elevation, that way if we improve we can reference the same throughout

File: lib/export/coldp/files/type_material.rb [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            host(o, co),                                           # host
+            date(co),                                           # date
```
Comment (mjy) [inline lib/export/coldp/files/type_material.rb:112]: dwc_event_date

**Pull request merged**